File: javascript/83.remove-duplicates-from-sorted-list.js
/*
 * @lc app=leetcode id=83 lang=javascript
 *
 * [83] Remove Duplicates from Sorted List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function (head) {
  if (head && head.next) {
    // Solution 1: one pointer
    // let p = head;
    // while (p && p.next) {
    //   if (p.val === p.next.val) p.next = p.next.next;
    //   else p = p.next;
    // }

    // Solution 2: slow and fast pointers
    let slow = head,
      fast = head;
    while (fast && fast.next) {
      fast = fast.next;
      if (slow.val !== fast.val) {
        slow.next = fast;
        slow = fast;
      }
    }
    slow.next = null;
  }

  return head;
};
// @lc code=end

File: rust/83.remove-duplicates-from-sorted-list.rs
/*
 * @lc app=leetcode id=83 lang=rust
 *
 * [83] Remove Duplicates from Sorted List
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut result = head.clone();
        let mut current = result.as_mut();
        while let Some(node) = current {
            let mut next = node.next.take();
            while let Some(n) = next.as_mut() {
                if n.val != node.val {
                    node.next = next;
                    break;
                }
                next = n.next.take();
            }
            current = node.next.as_mut();
        }
        return result;
    }
}
// @lc code=end
